File: Source/Extensions/Tessa.Extensions.Default.Client/Workflow/KrPermissions/KrKeepReadCardPermissionGetExtension.cs
using System;
using System.Threading.Tasks;
using Tessa.Cards.Extensions;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.UI;
using Tessa.UI.Cards;

namespace Tessa.Extensions.Default.Client.Workflow.KrPermissions
{
    public sealed class KrKeepReadCardPermissionGetExtension : CardGetExtension
    {
        public override Task BeforeRequest(ICardGetExtensionContext context)
        {
            ICardEditorModel editor;
            ICardModel model;
            if ((editor = UIContext.Current.CardEditor) != null
                && (model = editor.CardModel) != null
                && (editor.CurrentOperationType == CardEditorOperationType.SaveAndRefresh
                || editor.CurrentOperationType == CardEditorOperationType.Open))
            {
                if (KrToken.Contains(context.Request.Info))
                {
                    return Task.CompletedTask;
                }

                KrToken token = KrToken.TryGet(model.Card.Info);
                if (token != null && (token.CardID == context.Request.CardID || token.CardID == Guid.Empty))
                {
                    token.Set(context.Request.Info);
                }
            }

            return Task.CompletedTask;
        }
    }
}

File: WebClient SDK/src/default/workflow/krPermissions/krKeepReadCardPermissionGetExtension.ts
import { CardGetExtension, ICardGetExtensionContext } from 'tessa/cards/extensions';
import { UIContext } from 'tessa/ui';
import { ICardEditorModel, ICardModel, CardEditorOperationType } from 'tessa/ui/cards';
import { KrToken } from 'tessa/workflow';
import { Guid } from "tessa/platform";

export class KrKeepReadCardPermissionGetExtension extends CardGetExtension {

  public beforeRequest(context: ICardGetExtensionContext) {
    const uiContext = UIContext.current;
    let editor!: ICardEditorModel;
    let model!: ICardModel;
    if ((editor = uiContext.cardEditor!)
      && (model = editor.cardModel!)
      // Флаг сохраняем только при открытии при сохранении карточки, рефреш карточки скидывает режим редактирования
      && (editor.currentOperationType === CardEditorOperationType.SaveAndRefresh
        || editor.currentOperationType === CardEditorOperationType.Open)
    ) {
      if (KrToken.contains(context.request.info)) {
        return;
      }

      const token = KrToken.tryGet(model.card.info);
      if (token && (Guid.equals(token.cardId, context.request.cardId) || Guid.equals(token.cardId, Guid.empty))) {
        token.setInfo(context.request.info);
      }
    }
  }

}